x-pack/plugins/fleet/server/services/agent_policy.ts

```diff
@@ -531,6 +533,166 @@ class AgentPolicyService {
+    logger.info(`Starting bulk update of agent policies: ${agentPolicies.map(({ id }) => id)}`);
+
+    const existingAgentPolicies = await this.list(soClient, {
+      perPage: SO_SEARCH_LIMIT,
+      withPackagePolicies: true,
+    });
+
+    this.checkNameUniquenessBulk(agentPolicies, existingAgentPolicies.items);
+    this.checkPolicyExistsBulk(agentPolicies, existingAgentPolicies.items);
+
+    const policiesToUpdate: Array<Partial<AgentPolicy> & { id: string }> = [];
+    const failedPolicies: Array<{
+      agentPolicy: Partial<AgentPolicy>;
+      error: Error | SavedObjectError;
+    }> = [];
+
+    await pMap(agentPolicies, async (agentPolicy) => {
+      try {
+        this.checkTamperProtectionLicense(agentPolicy);
+        await this.checkForValidUninstallToken(agentPolicy, agentPolicy.id);
+        const existingAgentPolicy = existingAgentPolicies.items.find(
+          (i) => i.id === agentPolicy.id
+        );
+        if (!existingAgentPolicy) {
+          throw new AgentPolicyNotFoundError('Agent policy not found'); // This should never happen
+        }
+        if (agentPolicy?.is_protected && !policyHasEndpointSecurity(existingAgentPolicy)) {
+          logger.info(
+            'Agent policy requires Elastic Defend integration to set tamper protection to true'
+          );
+          // force agent policy to be false if elastic defend is not present
+          agentPolicy.is_protected = false;
+        }
+
+        if (existingAgentPolicy.is_managed && !options?.force) {
+          Object.entries(agentPolicy)
+            .filter(([key]) => !KEY_EDITABLE_FOR_MANAGED_POLICIES.includes(key))
+            .forEach(([key, val]) => {
+              if (!isEqual(existingAgentPolicy[key as keyof AgentPolicy], val)) {
+                throw new HostedAgentPolicyRestrictionRelatedError(`Cannot update ${key}`);
+              }
+            });
+        }
+
+        const { monitoring_enabled: monitoringEnabled } = agentPolicy;
+        const packagesToInstall = [];
+        if (!existingAgentPolicy.monitoring_enabled && monitoringEnabled?.length) {
+          packagesToInstall.push(FLEET_ELASTIC_AGENT_PACKAGE);
+        }
+        if (packagesToInstall.length > 0) {
+          await bulkInstallPackages({
+            savedObjectsClient: soClient,
+            esClient,
+            packagesToInstall,
+            spaceId: options?.spaceId || DEFAULT_SPACE_ID,
+            authorizationHeader: options?.authorizationHeader,
+          });
+        }
+        if (
+          existingAgentPolicy.status === agentPolicyStatuses.Inactive &&
+          agentPolicy.status !== agentPolicyStatuses.Active
+        ) {
+          throw new FleetError(
+            `Agent policy ${agentPolicy.id} cannot be updated because it is ${existingAgentPolicy.status}`
+          );
+        }
+
+        if (options.removeProtection) {
+          logger.info(`Setting tamper protection for Agent Policy ${agentPolicy.id} to false`);
+        }
```

Comment:
do we need to output so many messages to the logs (one per agent)? instead, can you consolidate them so that only one message is output with all of the agent policy ids listed?